Mathlib/RingTheory/Congruence/Hom.lean

```diff
@@ -0,0 +1,476 @@
+/-
+Copyright (c) 2025 Antoine Chambert-Loir. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Antoine Chambert-Loir
+-/
+
+import Mathlib.Algebra.Algebra.Subalgebra.Lattice
+import Mathlib.Algebra.Algebra.Subalgebra.Basic
+import Mathlib.Algebra.Algebra.Defs
+import Mathlib.Algebra.Group.Hom.Defs
+import Mathlib.RingTheory.Congruence.Basic
+import Mathlib.Algebra.Ring.Subsemiring.Basic
+import Mathlib.Algebra.Ring.Subring.Basic
+import Mathlib.Algebra.RingQuot
+
+/-!
+# Congruence relations and ring homomorphisms
+
+This file contains elementary definitions involving congruence
+relations and morphisms for rings and semirings
+
+## Main definitions
+
+* `RingCon.ker`: the kernel of a monoid homomorphism as a congruence relation
+* `RingCon.lift`, `RingCon.liftₐ`: the homomorphism / the algebra morphism
+  on the quotient given that the congruence is in the kernel
+* `RingCon.map`, `RingCon.mapₐ`: homomorphism / algebra morphism
+  from a smaller to a larger quotient
+
+* `RingCon.quotientKerEquivRangeS`, `RingCon.quotientKerEquivRange`,
+  `RingCon.quotientKerEquivRangeₐ` :
+  the first isomorphism theorem for semirings (using `RingHom.rangeS`),
+  rings (using `RingHom.range`) and algebras (using `AlgHom.range`).
+* `RingCon.comapQuotientEquivRangeS`, `RingCon.comapQuotientEquivRange`,
+  `RingCon.comapQuotientEquivRangeₐ` : the second isomorphism theorem
+  for semirings (using `RingHom.rangeS`), rings (using `RingHom.range`)
+  and algebras (using `AlgHom.range`).
+
+* `RingCon.quotientQuotientEquivQuotient`, `RingCon.quotientQuotientEquivQuotientₐ` :
+  the third isomorphism theorem for semirings (or rings) and algebras
+
+## Tags
+
+congruence, congruence relation, quotient, quotient by congruence relation, ring,
+quotient ring
+-/
+
+
+variable {M : Type*} {N : Type*} {P : Type*}
+
+open Function Setoid
+
+namespace RingCon
+
+section
+
+variable [NonAssocSemiring M] [NonAssocSemiring N] [NonAssocSemiring P] {c d : RingCon M}
+
+/-- The kernel of a ring homomorphism as a ring congruence relation. -/
+def ker (f : M →+* N) : RingCon M where
+  toSetoid := Setoid.ker f
+  add' h1 h2 := by
+    dsimp [Setoid.ker, onFun] at *
+    rw [map_add, h1, h2, map_add]
+  mul' h1 h2 := by
+    dsimp [Setoid.ker, onFun] at *
+    rw [map_mul, h1, h2, map_mul]
```

Comment:
```suggestion
  __ := comap ⊥ f
```
(not tested)
or just define the whole thing to be `comap ⊥ f`?

Reply:
Yes, that's even better!